inluxclient/query.go

```diff
@@ -0,0 +1,315 @@
+package influxclient
+
+import (
+	"encoding/base64"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// TableColumn holds flux query table column properties
+type TableColumn struct {
+	Name         string
+	DataType     string
+	IsGroup      bool
+	DefaultValue string
+}
+
+// String returns TableColumn string dump
+func (f *TableColumn) String() string {
+	return fmt.Sprintf("{name: %s, datatype: %s, defaultValue: %s, group: %v}", f.Name, f.DataType, f.DefaultValue, f.IsGroup)
+}
+
+const (
+	stringDatatype       = "string"
+	doubleDatatype       = "double"
+	boolDatatype         = "boolean"
+	longDatatype         = "long"
+	uLongDatatype        = "unsignedLong"
+	durationDatatype     = "duration"
+	base64BinaryDataType = "base64Binary"
+	timeDatatypeRFC      = "dateTime:RFC3339"
+	timeDatatypeRFCNano  = "dateTime:RFC3339Nano"
+)
+
+type parsingState int
+type parsingMode int
+
+const (
+	parsingStateNoTable parsingState = iota
+	parsingStateDataRow
+	parsingStateAnnotation
+	parsingStateNameRow
+	parsingStateError
+)
+
+const (
+	parsingModeTable parsingMode = iota
+	parsingModeRow
+)
+
+// QueryResults parses Flux query response stream.
+// Walking though the result set is done by repeatedly calling NextTable() and NextRow() until return false.
+// NextRow() can be also called initially, to advance straight to the first row of the first table.
+// Actual flux table schema (columns with names, data types, etc) is returned by Columns() method.
+// Data are acquired by Values() or by ValueByName() functions.
+// Preliminary end can be caused by an error, so when NextTable() or NextRow() return false, check Err() for an error.
+// Reader is automatically closed at the end reading or in case of an error.
+type QueryResults struct {
+	io.Closer
```

Comment:
I think I'd avoid embedding this in favour of an explicit `Close` method - that way we have the freedom to change instance variables on close if we want, and we don't really want it to be public.

Reply:
Public `Close()` can be useful when Flux CSV is not read till the end to close the HTTP response body.

Reply:
I'm not suggesting that you don't have a public `Close` method, but that it's defined explicitly rather than via embedding:
```
func (q *QueryResults) Close() error {
    return q.r.Close()
}
```

That way we can easily attach arbitrary logic to the Close method if needed, and there's no worry that a caller might change the `Closer` field of the struct.

Reply:
Ok